File: selected_coding_interview/codes/java/lc_8_string_to_integer_atoi/lc_8_string_to_integer_atoi_s1.java
/*
* File: lc_8_string_to_integer_atoi_s1.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_8_string_to_integer_atoi;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        public int myAtoi(String s) {
            char[] c = s.trim().toCharArray();
            if (c.length == 0) return 0;
            int res = 0, bndry = Integer.MAX_VALUE / 10;
            int i = 1, sign = 1;
            if (c[0] == '-') sign = -1;
            else if (c[0] != '+') i = 0;
            for (int j = i; j < c.length; j++) {
                if (c[j] < '0' || c[j] > '9') break;
                if (res > bndry || res == bndry && c[j] > '7') return sign == 1 ? Integer.MAX_VALUE : Integer.MIN_VALUE;
                res = res * 10 + (c[j] - '0');
            }
            return sign * res;
        }
    }

public class lc_8_string_to_integer_atoi_s1 {
    public static void main(String[] args) {
        // ======= Test Case =======
        // Test case 1
        String test_input = "example";

        // ====== Driver Code ======
        Solution slt = new Solution();
        var result = slt.myAtoi(test_input);
        System.out.println(result);

    }
}

File: selected_coding_interview/codes/cpp/lc_8_string_to_integer_atoi/lc_8_string_to_integer_atoi_s1.cpp
/*
 * File: lc_8_string_to_integer_atoi_s1.cpp
 * Created Time: 2025-12-30
 * Author: krahets
 */

#include "../include/include.hpp"

// ===== Solution Code =====
class Solution {
public:
    int myAtoi(string s) {
        int res = 0, bndry = INT_MAX / 10;
        int i = 0, sign = 1, length = s.size();
        if(length == 0) return 0;
        while(s[i] == ' ')
            if(++i == length) return 0;
        if(s[i] == '-') sign = -1;
        if(s[i] == '-' || s[i] == '+') i++;
        for(int j = i; j < length; j++) {
            if(s[j] < '0' || s[j] > '9') break;
            if(res > bndry || res == bndry && s[j] > '7')
                return sign == 1 ? INT_MAX : INT_MIN;
            res = res * 10 + (s[j] - '0');
        }
        return sign * res;
    }
};

int main() {
    // ======= Test Case =======
    // TODO: Add specific test case

    // ====== Driver Code ======
    Solution* slt = new Solution();
    // TODO: Call solution method and print result

    return 0;
}
